starboard/nplb/player_get_media_time_test.cc

```diff
@@ -0,0 +1,204 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "starboard/nplb/player_test_fixture.h"
+#include "starboard/string.h"
+#include "starboard/testing/fake_graphics_context_provider.h"
+#include "testing/gtest/include/gtest/gtest.h"
+
+namespace starboard {
+namespace nplb {
+namespace {
+
+using ::testing::ValuesIn;
+
+typedef SbPlayerTestFixture::GroupedSamples GroupedSamples;
+typedef testing::FakeGraphicsContextProvider FakeGraphicsContextProvider;
+
+class SbPlayerGetMediaTimeTest
+    : public ::testing::TestWithParam<SbPlayerTestConfig> {
+ protected:
+  FakeGraphicsContextProvider fake_graphics_context_provider_;
+};
+
+TEST_P(SbPlayerGetMediaTimeTest, SunnyDay) {
+  const int64_t kDurationToPlay = 1'000'000;  // 1 second
+
+  SbPlayerTestFixture player_fixture(GetParam(),
+                                     &fake_graphics_context_provider_);
+  if (HasFatalFailure()) {
+    return;
+  }
+
+  GroupedSamples samples;
+  if (player_fixture.HasAudio()) {
+    samples.AddAudioSamples(
+        0, player_fixture.ConvertDurationToAudioBufferCount(kDurationToPlay));
+    samples.AddAudioEOS();
+  }
+  if (player_fixture.HasVideo()) {
+    samples.AddVideoSamples(
+        0, player_fixture.ConvertDurationToVideoBufferCount(kDurationToPlay));
+    samples.AddVideoEOS();
+  }
+  ASSERT_NO_FATAL_FAILURE(player_fixture.Write(samples));
+  ASSERT_NO_FATAL_FAILURE(player_fixture.WaitForPlayerPresenting());
+
+  int64_t start_system_time = CurrentMonotonicTime();
+  int64_t start_media_time = player_fixture.GetCurrentMediaTime();
+
+  ASSERT_NO_FATAL_FAILURE(player_fixture.WaitForPlayerEndOfStream());
+
+  int64_t end_system_time = CurrentMonotonicTime();
+  int64_t end_media_time = player_fixture.GetCurrentMediaTime();
+
+  const int64_t kDurationDifferenceAllowance = 500'000;  // 500 ms
+  EXPECT_NEAR(end_media_time, kDurationToPlay, kDurationDifferenceAllowance);
+  EXPECT_NEAR(end_system_time - start_system_time + start_media_time,
+              kDurationToPlay, kDurationDifferenceAllowance);
+
+  SB_DLOG(INFO) << "The expected media time should be " << kDurationToPlay
+                << ", the actual media time is " << end_media_time
+                << ", with difference "
+                << std::abs(end_media_time - kDurationToPlay) << ".";
+  SB_DLOG(INFO) << "The expected total playing time should be "
+                << kDurationToPlay << ", the actual playing time is "
+                << end_system_time - start_system_time + start_media_time
+                << ", with difference "
+                << std::abs(end_system_time - start_system_time +
+                            start_media_time - kDurationToPlay)
+                << ".";
+}
+
+TEST_P(SbPlayerGetMediaTimeTest, NotPlayWithInput) {
```

Comment:
We can try get media time before writing any inputs. Or if we want to test media time while video is paused, we need to explicitly pause or set the playback rate to 0.

Reply:
Fixed.